### ISSUE TITLE
Warning about reference target not being found for container types

### ISSUE THREAD

**Issue by tjanez** [sphinx-doc/sphinx#3320]
### Problem
Sphinx 1.5 introduced [container types](http://www.sphinx-doc.org/en/1.5.1/domains.html#info-field-lists), e.g.:
```
:type priorities: list(int)
:type priorities: list[int]
:type mapping: dict(str, int)
:type mapping: dict[str, int]
:type point: tuple(float, float)
:type point: tuple[float, float]
```

When I use them, the rendered documentation looks ok, but `sphinx-build` returns a warning about reference target not being found for the container type.

#### Procedure to reproduce the problem
Here is a minimal example to reproduce the problem:
- prepare a new directory for a testing project
- create `conf.py` file with the following contents:

``` python
master_doc = 'index'

project = 'Foo'
version = '0.1'
release = '0.1'
author = 'Bar'
copyright = 'Baz'

extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.intersphinx',
]

intersphinx_mapping = {
    'python': ('https://docs.python.org/3', None),
}

# Warn about all references where the target cannot be found
nitpicky = True

# Add current directory to Python Path
import os
import sys
sys.path.insert(0, os.getcwd())
```
- create `index.rst` file with the following contents:

``` rst
.. automodule:: foo
  :members:
```

- create `foo.py` with the following contents:

```python
def get_point(x, y):
    """Convert given coordinates to a point.

    :param float x: x-coordinate
    :param float y: y-coordinate

    :return: (x-coordinate, y-coordinate)
    :rtype: tuple(float, float)
    """
    return (1.3, 2.5)
```
- build the documentation with: `sphinx-build -E ./ html/` (ensure `html` directory exists)

#### Error logs / results

Sphinx gives the following error for the container type:
`foo.py:docstring of foo.get_point:None: WARNING: py:obj reference target not found: `

Here is the full log of the `sphinx-build` run:
```
Running Sphinx v1.5.1
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index                                                                                                             
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index                                                                                                              
/home/tadej/Genialis/Temp/sphinx-container-type-problem/foo.py:docstring of foo.get_point:None: WARNING: py:obj reference target not found: 
generating indices... genindex py-modindex
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 1 warning.
```

#### Expected results
No warning is emitted.

### Environment info
- OS: Fedora 24
- Python version: 3.5.2
- Sphinx version: 1.5.1

**Comment by lazka** [comment]
Still a problem with 1.5.5

**Comment by tk0miya** [comment]
Thank you for reporting!
Fixed at #3744.
